### ISSUE TITLE
Let's implement threads.mutex

### ISSUE THREAD

**Issue by levBagryansky** [objectionary/eo-threads#12]
@yegor256 Let's implement `mutex` object. It will have `mutex.lock` and `mutex.unlock` attributes so only one `thread` can own the `mutex` in a moment.

**Comment by levBagryansky** [comment]
@yegor256 Can I do it?

**Comment by yegor256** [comment]
@levBagryansky sure, but I would name them `acquire` and `release`:

```
QQ.threads.mutex >
m.acquire > a
a.release
```

This is the most logical way of dealing with locks, I believe

**Comment by levBagryansky** [comment]
@yegor256 ok. I thought `lock` and `unlock` should return `TRUE`. So we would use it like
```
[] > app
  QQ.threads.mutex > m
  seq > @
    m.acquire
    m.release
```
But do `m` and `a` refer to the sane object in your sample?

**Comment by levBagryansky** [comment]
@yegor256 Will we use `mutex` in this way:
```
[] > app
  QQ.threads.mutex > m
  seq > @
    m.acquire
    do-smth
    m.release
```
?

**Comment by yegor256** [comment]
@levBagryansky in this case we can only acquire ONE lock, while a proper mutex should allow multiple of them:

```
[] > app
  QQ.threads.mutex 2 > m
  seq > @
    m.acquire > first
    do-smth
    m.acquire > second
    first.release
    second.release
```

**Comment by levBagryansky** [comment]
How can we lock the already existing `mutex` if `acquire` is a new `mutex`? Can we do?:
[] > app
  mutex > m
  seq 
  QQ.threads.thread

**Comment by levBagryansky** [comment]
@yegor256 may be we can write 
```
[] > app
  QQ.threads.mutex' > m
```
to create a new mutex?

**Comment by yegor256** [comment]
@levBagryansky our `mutex` should be very similar to https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/Semaphore.html But it would be better to have "personalized" locks: if you acquire one, you must return exactly the same one. That's why:

```
mutex 100 > m
m.acquire > a
a.release  # instead of m.release
```

**Comment by levBagryansky** [comment]
@yegor256 So when in the same time another thread will do 
`m.acquire > b` will it wait for `a.release`?

**Comment by yegor256** [comment]
@levBagryansky yes, of course

**Comment by levBagryansky** [comment]
@yegor256 ok thanks

**Comment by levBagryansky** [comment]
@yegor256 Must the second `a.release` be legal or cause an `error`?

**Comment by yegor256** [comment]
@levBagryansky must cause and error, of course

**Comment by yegor256** [comment]
> But when it tries to dataize a.release it go to a and again acquires mutex to return lock without its dataization

I don't think it's true. In order to dataize `a.release` EO doesn't need to dataize `a.@`

**Comment by levBagryansky** [comment]
@yegor256 It does not dataize `a.@` but is dataize `m.acquire` again. I can demonstrate it.
This is simplified mutex implementation:
```
[] > mutex
  [] > acquire
    seq > @
      QQ.io.stdout "Acquire\n"
      lock

  [] > lock
    TRUE > @
    [] > release
      TRUE > @
```
This is the test:
```
[] > twice-lock
  mutex > m
  seq > @
    m.acquire > a
    a.release
``` 
 When we run the test it ptints "Acquire" twice that poves it would acqure the mutex twice:
```
[INFO] Running EOorg.EOeolang.EOthreads.EOtwice_lockTest
Acquire
Acquire
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.021 s - in EOorg.EOeolang.EOthreads.EOtwice_lockTest
```

**Comment by 0pdd** [comment]
@levBagryansky 2 puzzles [#33](https://github.com/objectionary/eo-threads/issues/33), [#34](https://github.com/objectionary/eo-threads/issues/34) are still not solved.

**Comment by 0pdd** [comment]
@levBagryansky 5 puzzles [#33](https://github.com/objectionary/eo-threads/issues/33), [#34](https://github.com/objectionary/eo-threads/issues/34), [#51](https://github.com/objectionary/eo-threads/issues/51), [#52](https://github.com/objectionary/eo-threads/issues/52), [#53](https://github.com/objectionary/eo-threads/issues/53) are still not solved.

**Comment by 0pdd** [comment]
@levBagryansky 4 puzzles [#33](https://github.com/objectionary/eo-threads/issues/33), [#34](https://github.com/objectionary/eo-threads/issues/34), [#52](https://github.com/objectionary/eo-threads/issues/52), [#53](https://github.com/objectionary/eo-threads/issues/53) are still not solved; solved: [#51](https://github.com/objectionary/eo-threads/issues/51).

**Comment by 0pdd** [comment]
@levBagryansky 3 puzzles [#33](https://github.com/objectionary/eo-threads/issues/33), [#34](https://github.com/objectionary/eo-threads/issues/34), [#52](https://github.com/objectionary/eo-threads/issues/52) are still not solved; solved: [#51](https://github.com/objectionary/eo-threads/issues/51), [#53](https://github.com/objectionary/eo-threads/issues/53).

**Comment by 0pdd** [comment]
@levBagryansky 2 puzzles [#33](https://github.com/objectionary/eo-threads/issues/33), [#52](https://github.com/objectionary/eo-threads/issues/52) are still not solved; solved: [#34](https://github.com/objectionary/eo-threads/issues/34), [#51](https://github.com/objectionary/eo-threads/issues/51), [#53](https://github.com/objectionary/eo-threads/issues/53).

**Comment by 0pdd** [comment]
@levBagryansky the puzzle [#33](https://github.com/objectionary/eo-threads/issues/33) is still not solved; solved: [#34](https://github.com/objectionary/eo-threads/issues/34), [#51](https://github.com/objectionary/eo-threads/issues/51), [#52](https://github.com/objectionary/eo-threads/issues/52), [#53](https://github.com/objectionary/eo-threads/issues/53).

**Comment by 0pdd** [comment]
@levBagryansky all 5 puzzles are solved here: [#33](https://github.com/objectionary/eo-threads/issues/33), [#34](https://github.com/objectionary/eo-threads/issues/34), [#51](https://github.com/objectionary/eo-threads/issues/51), [#52](https://github.com/objectionary/eo-threads/issues/52), [#53](https://github.com/objectionary/eo-threads/issues/53).
